### ISSUE TITLE
Feature request: report missing Qed. somehow

### ISSUE THREAD

**Issue by Blaisorblade** [ProofGeneral/PG#154]
With Coq 8.5, ProofGeneral appears to accept the following script (minimized from a 1000-line proof script).

```
Lemma a1 : 1 + 1 = 2.
auto.

Lemma a2 : 1 + 2 = 3.

Lemma a3 : 1 + 3 = 4.
  auto.
Qed.
```

However, `coqc` rejects it:
```
$ coqc Foo.v
Error: There are pending proofs
```

@JasonGross mentioned that the pending proofs can be discovered with `End` (and `End a3` indeed shows the missing proofs), or `Show Proof.` (but that only shows `?Goal` with Coq 8.5).

But at least I (a Coq newbie) can't honestly tell from the ProofGeneral interface — when I step to the end of the script, it shows an empty `*goals*` window, at least sometimes.
So much that the 1000-line script got committed by our resident Coq expert (not me).

OTOH, if I resume the outer proofs with `auto. Qed. Qed.` I get a warning:

> Warning: Nested proofs are deprecated and will stop working in the next Coq version

suggesting the problem might fix itself soon (though this is debated https://sympa.inria.fr/sympa/arc/coq-club/2015-08/msg00031.html).

Quoting @JasonGross wrote that https://github.com/ProofGeneral/PG/issues/37#issuecomment-281146280:
> Report a separate issue for the Qed; the comment issue described here is a bug in PG; PG parses the document differently from coqtop and coqc, and this causes unexpected behavior. Your issue is that PG doesn't tell you if your file is unfinished (has unterminated proofs or open sections), but coqtop acts the same way as PG here; it's not clear when or how PG should give you this information (and you can discover it at any time by seeing what happens if you do Show Proof. or End foo.; no need to bisect), so it's more properly a feature request. I'd actually suggest making a feature request on the Coq bug tracker, asking for the error message to tell you what line the proof started on, and possibly the theorem name.

**Comment by Blaisorblade** [comment]
>  I'd actually suggest making a feature request on the Coq bug tracker, asking for the error message to tell you what line the proof started on, and possibly the theorem name.

I found this reported as https://coq.inria.fr/bugs/show_bug.cgi?id=5275.

**Comment by Matafou** [comment]
Hi,
To be honest I don't see how and why Proofgeneral should complain about an
unterminated proof. After all most of the time I do have an unterminated
proof at the end of my buffer, that is the whole point of proof interfaces:
deal with unterminated proofs and help the user to finish them. coqc is for
finished proofs, coqtop and ides are for unfinished proofs. you may check
the messsage returnd by  "Show." to check by hand if there is a pending
proof (anything else than "there is no pending proof" means there is at
least one pending pooof).

At what time should an unterminated proof be detected? When another proof
is started? Then that would basically forbid nested proofs. Since Coq
supports them for now I am not convinced we should forbid them. If and once
Coq forbud them then PG will also.

Remark: If you hit (indent-region) on the whole file (this is slow on big
files) you should see the place where the proof is not finished (things
will start to be shifted o, the right there).


Best regards,
Pierre


2017-02-20 20:59 GMT+01:00 Paolo G. Giarrusso <notifications@github.com>:

**Comment by Blaisorblade** [comment]
> Hi,
> To be honest I don't see how and why Proofgeneral should complain about an
> unterminated proof. After all most of the time I do have an unterminated
> proof at the end of my buffer, that is the whole point of proof interfaces:

Sorry I was unclear (and unsure of what to do) — I don't suggest ProofGeneral should forbid nested proofs, it'd be enough if the `*goals*` window were up-to-date (as it usually is), so I wouldn't need to learn I might need to update it and how to do that.
In years (of very light usage, I use Agda more often) I had never learned the point of the leftmost eyes button, I probably concluded quickly I never needed it and stopped looking at it.

`*goals*` can have either too few or too many (!) goals (maybe in relation to #103).
- I figured jumping to the end of a `Qed.` leaves `*goals*` with the same content it had at the start, even when proofs are still open. That's the confusing point of the interface. Confusing enough that a more experienced user committed this file.

As a consequence of that rule, I got `*goals*` to have content when no goal is actually open, and there no refresh helps:
- Add `auto. Qed. Qed.` at the end
- Jump right after `Lemma a1 : 1 + 1 = 2.`. Then, `*goals*` shows
```
1 subgoal, subgoal 1 (ID 2)

  ============================
   1 + 1 = 2

```
- Jump at the end of the file. `*goals*` has the same content, and neither `C-c C-l` nor the leftmost "eyes" button update it. `Show.`, `Show Proof.`, `End` simply fail, as there's no open proof, which might be correct but still doesn't update `*goals*`.

**Comment by Matafou** [comment]
I see. Indeed this is an old bug but it is the first time I see a
description of it being really annoying.
I will try to have a look at it (even if we may switch to xml/pg some day).


P.

2017-02-21 15:24 GMT+01:00 Paolo G. Giarrusso <notifications@github.com>:

**Comment by Blaisorblade** [comment]
> Indeed this is an old bug but it is the first time I see a
> description of it being really annoying.
> I will try to have a look at it (even if we may switch to xml/pg some day).

Thanks! It's clearly not the topmost priority, especially if xml/pg is the direction long term. I'm just saying in this admittedly rare scenario it can be confusing. Especially if the Coq expert who can deal with this is in another country.

BTW, kudos to PG's usability otherwise. In particular, "Compile before load" is much more usable than alternatives I've tried to learn. And that's why I can bother reporting these small bugs to make it even better!

And thanks for offering it for free even to users who don't have a chance to contribute! :-)

**Comment by Matafou** [comment]
Compile Before Require was made by Hendriks Tews and indeed it is the best feature PG got in years imho.

**Comment by Matafou** [comment]
For the record the fix I propose in PR #156 is to perform a Show when closing one proof while there is another open proof.
Therefore you will get either a subgoal printed or "No more subgoals,".

**Comment by Blaisorblade** [comment]
Closing after #156 was merged (haven't personally retested, shouldn't be necessary) :-)
